Services/RefundService.php

```diff
@@ -0,0 +1,281 @@
+                    $this->updateRefundedItemsOnOrderDetail($detail, $detail->getQuantity());
+                }
+            }
+        }
+
+        return $refund;
+    }
+
+    /**
+     * Refund a Mollie order
+     *
+     * @param Order       $order
+     * @param MollieOrder $mollieOrder
+     *
+     * @return bool|Refund
+     * @throws \Exception
+     *
+     */
+    public function partialRefundOrder(
+        Order $order,
+        Detail $detail,
+        MollieOrder $mollieOrder,
+        OrderLine $orderLine,
+        $quantity = 1
+    ) {
+        $apiClient = $this->getMollieApi($order->getShop()->getId());
+
+        if (empty($this->modelManager) || $apiClient === null) {
+            return false;
+        }
+
+        $data = [
+            'lines' => [
+                [
+                    'id' => $orderLine->id,
+                    'quantity' => $quantity,
+                ],
+            ]
+        ];
+
+        /** @var Refund $refund */
+        $refund = $apiClient->orderRefunds->createFor($mollieOrder, $data);
+
+        if ($refund !== null) {
+            $this->processRefund($order);
+            $this->updateRefundedItemsOnOrderDetail($detail, $quantity);
+        }
+
+        return $refund;
+    }
+
+    /**
+     * Refund a Mollie payment
+     *
+     * @param Order         $order
+     * @param MolliePayment $molliePayment
+     *
+     * @return Mollie\Api\Resources\BaseResource
+     *
+     * @throws \Exception
+     */
+    public function refundPayment(Order $order, MolliePayment $molliePayment)
+    {
+        $refund = $molliePayment->refund(
+            [
+                'amount' => [
+                    'currency' => $order->getCurrency(),
+                    'value' => number_format($order->getInvoiceAmount(), 2, '.', '')
+                ]
+            ]
+        );
+
+        if ($refund !== null) {
+            $this->processRefund($order);
+        }
+
+        return $refund;
+    }
+
+    /**
+     * Refund a Mollie payment
+     *
+     * @param Order         $order
+     * @param MolliePayment $molliePayment
+     * @param float         $amountToRefund
+     *
+     * @return Mollie\Api\Resources\BaseResource
+     *
+     * @throws \Exception
+     */
+    public function partialRefundPayment(Order $order, MolliePayment $molliePayment, $amountToRefund)
+    {
+        $refund = $molliePayment->refund([
+            'amount' => [
+                'currency' => $order->getCurrency(),
+                'value' => number_format($amountToRefund, 2, '.', '')
+            ]
+        ]);
+
+        if ($refund !== null) {
+            $this->processRefund($order);
+        }
+
+        return $refund;
+    }
+
+    /**
+     * Send the status e-mail
+     *
+     * @param Order $order
+     *
+     * @return bool
+     *
+     * @throws \Exception
+     */
+    public function processRefund(Order $order)
+    {
+        if (empty($this->config) || empty($this->modelManager)) {
```

Comment:
if it works with fail-fast just remove it, or add an exception if the if is required, thx